### ISSUE TITLE
请问有在qwen 1.5B或其他小参数量的llm上做过实验吗？

### ISSUE THREAD

**Issue by bigcash** [MooreThreads/MooER#24]
实验效果如何？
我使用Qwen2.5-1.5B-Instruct来lora微调，或者全参微调，模型都不收敛。当然，在现有的7B的模型上继续做微调，是没有问题的。

炼丹水平还是不够~

**Comment by lzl-mt** [comment]
你的微调是针对什么任务呢？ASR、AST、S2ST还是Omni？我们在0.5B和1.5B的模型做过语音翻译的实验，效果会变差，但是不会出现没法收敛的情况。

**Comment by bigcash** [comment]
微调的是ASR任务，标注中增加了标点。在Qwen2.5-1.5B-Instruct的模型上做的lora，跑了一个完整的epoch下来，日志中最后的acc也不超过0.5。现在在跑全量的微调（但embed_tokens冻结），跑了大概有30w条数据了，loss和acc还是上下震动的厉害，如下：

2024-11-19 11:10:24 | INFO | root | Training Epoch: 1/1, step 102656 lr 4.638939775257248e-05 completed (loss: 4.192238807678223, acc: 0.25)
2024-11-19 11:12:47 | INFO | root | Training Epoch: 1/1, step 102912 lr 4.638939775257248e-05 completed (loss: 4.908051013946533, acc: 0.2631579041481018)
2024-11-19 11:15:12 | INFO | root | Training Epoch: 1/1, step 103168 lr 4.638939775257248e-05 completed (loss: 1.650698184967041, acc: 0.5)
2024-11-19 11:17:36 | INFO | root | Training Epoch: 1/1, step 103424 lr 4.648963071619275e-05 completed (loss: 3.48872971534729, acc: 0.5)
2024-11-19 11:19:59 | INFO | root | Training Epoch: 1/1, step 103680 lr 4.648963071619275e-05 completed (loss: 4.081971645355225, acc: 0.5)
2024-11-19 11:22:22 | INFO | root | Training Epoch: 1/1, step 103936 lr 4.648963071619275e-05 completed (loss: 2.7896568775177, acc: 0.5714285969734192)
2024-11-19 11:24:47 | INFO | root | Training Epoch: 1/1, step 104192 lr 4.648963071619275e-05 completed (loss: 4.351863384246826, acc: 0.3333333432674408)
2024-11-19 11:27:11 | INFO | root | Training Epoch: 1/1, step 104448 lr 4.6588876146874495e-05 completed (loss: 3.4536798000335693, acc: 0.2857142984867096)
2024-11-19 11:29:35 | INFO | root | Training Epoch: 1/1, step 104704 lr 4.6588876146874495e-05 completed (loss: 4.596532344818115, acc: 0.2857142984867096)

**Comment by lzl-mt** [comment]
建议不要SFT，会影响LLM本身学好的能力。如果你新增加了token，那么Embedding层和llmhead做sft，其他做lora；另外ASR的PUNC任务是基于一个训好的ASR模型去进行的吗，还是从裸的LLM开始训的呢

**Comment by bigcash** [comment]
”建议不要SFT，会影响LLM本身学好的能力。“ 这个意思是不要做llm的全量微调吗？

”如果你新增加了token，那么Embedding层和llmhead做sft，其他做lora；“ 没有增加token，仅多了paraformer encoder的语音token embedding（和本项目的思路一致），这里冻结了embed_tokens，是冻结了文字onehot的token到embedding部分，感觉这里的参数不应该更新，llm在微调过程中只需要适应新增加的语音token embedding，进而最后自回归输出正确的label。

ASR里增加punc任务（label来自开源数据集自带的标点，如果没有标点，则使用iic/punc_ct-transformer_cn-en-common-vocab471067-large模型做伪标注），做了两次实验：
第一，使用本项目的MooER-80K-v2模型，做lora微调，数据集有开源+内部。这个之前微调没问题，loss收敛。
第二，使用了Qwen2.5-1.5B-Instruct模型，分别做了lora和全参微调，这时projector的参数是随机初始化的，所以相当于从裸的LLM开始从头训的。数据集仅使用开源数据（想着做个二阶段微调，等模型收敛了在第2阶段上开源+内部数据的，做进一步的微调，结果第1步就卡住了），这里的开源数据就是gigaspeech+wenetspeech+common voice+multi cn，大概1.6w小时。

**Comment by lzl-mt** [comment]
是的，我们实验过对于0.5B和1.5B的模型来说，在AST的任务上，lora的效果会显著优于sft，但原因也是我们的数据是脏的数据（伪标签）。
我训练过程中没有遇到loss不收敛的问题，你的lora的配置和我们提供的是一样的么？包括Adapter的下采样率等。

**Comment by bigcash** [comment]
lora的配置有变化，micro batch=8, gpu num=3, gradient_accumulation_steps=32，lr=1e-4, warmup_num_steps=1000, lora_rank=128, lora_alpha=16, 样本量=15M条, num_epochs=1, total_num_steps=100000，显卡是3张24G的3090

Adapter的下采样率是一样的，也是2倍，只是因为1.5B的维度是1536，所以llm_dim=1536

**Comment by bigcash** [comment]
punc数据，英文不再是全大写，改为首字母大写了

**Comment by lzl-mt** [comment]
只能给几个建议了...
1. LLM不动，只Finetune Adapter，看看loss是否能下降
2. 如果1能成，lora的rank变小试试
3. encoder是否参与了训练？建议encoder frozen住

**Comment by bigcash** [comment]
好的，我尝试一下，多谢啦！

**Comment by MonolithFoundation** [comment]
会开源更小尺寸的模型吗？for speech translation 任务

**Comment by lzl-mt** [comment]
> 会开源更小尺寸的模型吗？for speech translation 任务

我们可以计划短期内开源一个1.5B的AST的模型

**Comment by MonolithFoundation** [comment]
这个内部已经有指标了吗，效果大概差多少

**Comment by lzl-mt** [comment]
> 这个内部已经有指标了吗，效果大概差多少

Covost2 Zh2en的Bleu大概在 21.X

**Comment by MonolithFoundation** [comment]
会开园吗？

**Comment by lzl-mt** [comment]
> 会开园吗？

抱歉，最近太忙了，我们下周内开源1.5B的模型

**Comment by rookie0607** [comment]
> 好的，我尝试一下，多谢啦！

问题解决了吗？ @bigcash

**Comment by yiliu-mt** [comment]
Hi, we have uploaded the smaller model in ModelScope. Hope that helps!
https://modelscope.cn/models/MooreThreadsSpeech/MooER-MTL-5K-1.5B

**Comment by MonolithFoundation** [comment]
Hi, the 5K model not very good enough compare with paraformer. In Chinese ASR usage, the value is not very high.

**Comment by bigcash** [comment]
> > 好的，我尝试一下，多谢啦！
> 
> 问题解决了吗？ @bigcash

抱歉刚看到，问题解决了。
我这里是因为我对mooer_model.py中的line36
```python
ckpt_path = model_config.get("adapter_path", "")
```
做了修改，我原以为这里只是adapter_path相关的东西，所以将其改为了新的变量，然后加了一段代码来特意加载adapter模型的参数。因为我修改了adapter模型的结构（使用稍小llm时dim变小了），想使用原模型部分的参数，但这么做导致了这个问题，所以我将处理步骤改为使用原有的代码，将参数重用改为单独手动复用，然后保存一个新的模型文件。训练时直接加载这个模型参数文件开始训练，就没问题了。
